File: algorithms/javaalgo/src/test/java/leetcode/number201to300/number201to210/number206/ReverseLinkedListTest.java
package leetcode.number201to300.number201to210.number206;

import leetcode.util.ListNode;

public class ReverseLinkedListTest {

    public ListNode reverseList(ListNode head) {
        return reverse(head, null);
    }

    private ListNode reverse(ListNode node, ListNode prev) {
        if (node == null) {
            return prev;
        }
        ListNode next;
        next = node.next;
        node.next = prev;
        return reverse(next, node);
    }
}

File: algorithms/javascriptalgo/leetcode/number201-300/number201-210/number206/reverseLinkedList.test.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const reverseList = function(head) {
  const reverse = (node, prev) => {
    if (!node) return prev;

    let next = node.next;
    node.next = prev;
    return reverse(next, node);
  }

  return reverse(head, null);
};
